### DISCUSSION TITLE
Game launcher for seamless streaming experience using AutoHotKey

### DISCUSSION THREAD

**Discussion by basvd** [LizardByte/community#424] [Sunshine]
A problem I've experienced with Moonlight/Sunshine is that the stream session stays active even after a game is finished. This is mainly the case when launching games through Steam and similar launchers. Sunshine is not able to figure out which process to join for the duration of the session.

I have worked around this problem using AutoHotKey by writing a script that runs the game and then waits for a specified executable to start. The script keeps running until that process stops. Since Sunshine is watching the script, the streaming session will stop as it should when the game ends.

I'm pretty sure this could be implemented in Sunshine, however I do not know how to make it cross-platform. My script is limited to Windows due to AutoHotkey.

**TLDR**

Here's a rundown of the features:
- Launches a game and ends the stream when the game stops
- Hides the desktop behind a black screen with Sunshine logo
- Hides the mouse cursor while the game launches
- Should work with any type of launcher or games directly

Installation and usage:
- Get AutoHotkey if you don't have it already: https://www.autohotkey.com/
- Get the script here: https://gist.github.com/basvd/6549932f3ded2f26ab82d665c42d74ca
- Save this [256x256 Sunshine logo](https://github.com/LizardByte/Sunshine/assets/194770/75881045-2fe4-4a46-a389-c252eb4f4a20)  as `sunshine.png` in the same directory as the script.
- Use AutoHotKey to convert the .ahk script into a .exe (optional, but makes usage easier).

Now you can configure an application to launch using a command like:
```
"C:\Users\Bas\Local apps\SunshineLauncher.exe" steam://rungameid/374320 DarkSoulsIII.exe
```
Both "Continue streaming..." options should be unchecked.

Note that you have to provide the name of an executable as well. This should be the main .exe for the game, which you can usually find in the game installation folder or by inspecting the task manager when the game is running.

Small caveat: There is a chance this script doesn't work well with games that run in borderless windowed mode, this is because the overlay is set to AlwaysOnTop. You can remove this from the script if it's a problem, but it will make the taskbar visible during launch.

**Comment by thetommah** [comment]
I have tested this and it does look great, I have some experience with AHK as well and the script seems pretty straight forward and I like it.  I have run into 1 issue with Baldur's gate 3 though, after 10ish seconds it just ends the stream (game still running on host).  Any idea what would cause that?  The exe is correct (bg3.exe) and the script works great on other steam games like Helldivers 2.  Just wondering if you had any idea because I can't figure it out.

**Reply by basvd** [reply]
I don't have BG3 so I can't check this for myself. One possibility is that the bg3.exe takes longer than 10 seconds to launch, maybe because the game has a launcher of its own. In that case, you can try increasing the timeout on line 25 of the script (i.e. the 10 in `ProcessWait(exeName, 10)`).
Another possibility is that bg3.exe does launch before the timeout, but then launches itself again while stopping the first process. If this is the case, then it will be tricky to get it to work. I think the first option is more likely though.

If you want to debug this, edit the script to write the value of `pid` to a file using [FileAppend](https://www.autohotkey.com/docs/v2/lib/FileAppend.htm). Then, while the game is running, use the task manager to inspect the actual PID of the bg3.exe process. If the written PID is zero, then it has to do with the ProcessWait timeout and increasing it should fix it. If the PIDs are different and non-zero, then it's probably due to the second option.

[a comment by an automated account is not shown]
